Dynamic friction

EstherHeck opened the pull request:
Pull Request Checklist. Please read and check each box with an X. Delete any part not applicable. Ask on the [forum](https://community.geodynamics.org/c/aspect) if you need help with any step.

*Describe what you did in this PR and why you did it.*

### Before your first pull request:

* [x] I have read the guidelines in our [CONTRIBUTING.md](../blob/master/CONTRIBUTING.md) document.

### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../blob/master/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [ ] I have tested my new feature locally to ensure it is correct.
* [ ] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/master/tests/) directory.
* [ ] I have added a changelog entry in the [doc/modules/changes](../blob/master/doc/modules/changes) directory that will inform other users of my change.

gassmoeller committed:
Add a benchmark that solves the entropy equation

MFraters committed:
Add input file.

MFraters committed:
add required plugin.

MFraters committed:
update readme.

Commit:
Change input method, allowing single input value for compositions and phases

Commit:
took visco_plastic files from dynamic_friction branch into branch newly creasted from master to avoid rebase troubles

gassmoeller committed:
Make script standard compliant

Commit:
changed smoothness factor to smoothness exponent

anne-glerum committed:
Allow add tang BI for zero vel BI

gassmoeller committed:
Suggest new particle property interface

MFraters committed:
separate twomergedboxes from box geometry model and define a independent height_above_reference_surface function.

MFraters committed:
Fix finite difference in visco plastic material model.

MFraters committed:
Fix visco plastic derivatives tests and their results.

MFraters committed:
Add visco plastic newton nonlinear channelflow test.

MFraters committed:
indenting.

MFraters committed:
apply diff patch.

MFraters committed:
add more tests for reference before optimizing.

MFraters committed:
do no rebuild (newton) stokes matrix when (hopefully) not needed.

Commit:
add Q1 viscosity for GMG

danieldouglas92 committed:
Add viscoelastic cantilever benchmark

anne-glerum committed:
Add test

danieldouglas92 committed:
Create 20200513_douglas

danieldouglas92 committed:
Include references to analytic solution

danieldouglas92 committed:
Update input file for a 20km thick beam

danieldouglas92 committed:
Addressed comments

MFraters committed:
Apply suggestions from code review

Co-authored-by: Rene Gassmoeller <rene.gassmoeller@mailbox.org>

Commit:
add IDR(s) solver

MFraters committed:
astyle comments from rene.

anne-glerum committed:
Add info unit boundary velocity

anne-glerum committed:
Fix compiler warning

anne-glerum committed:
Bypass topography in coordinate conversion

gassmoeller committed:
Merge pull request #3508 from anne-glerum/fix_compiler_warning_adiabatic_boundary

Fix compiler warning

gassmoeller committed:
Merge pull request #3489 from MFraters/fix_visco_plastic_finite_difference

Fix visco plastic finite difference

gassmoeller committed:
Merge pull request #3480 from anne-glerum/allow_displacement_on_add_tang_BI_zero_vel

Allow tangential mesh velocities for zero Stokes velocity boundaries

pkestene committed:
fixing a small typo in documentation

gassmoeller committed:
Merge pull request #3507 from anne-glerum/comment_mesh_def_function

Add info unit boundary velocity

tjhei committed:
Merge pull request #3513 from pkestene/patch-1

fixing a small typo in documentation

Commit:
include ghost cells

Commit:
compress

gassmoeller committed:
Merge pull request #3514 from tcclevenger/fix_tangential_bc_gmg

Fix tangential BCs for GMG

gassmoeller committed:
Merge pull request #3499 from gassmoeller/solve_entropy_equation

Add a benchmark that solves the entropy equation

tjhei committed:
update parameters

tjhei committed:
fix two small doc issues

- missing bib reference
- missing parameter documentation

gassmoeller committed:
Merge pull request #3516 from tjhei/doc-fixes

fix two small doc issues

gassmoeller committed:
Merge pull request #3515 from tjhei/maintenance_jun2020

maintenance June 2020

tjhei committed:
fix parameter_view in firefox

the fancy logic to call the autorun function doesn't work in firefox.
Fix this.

jdannberg committed:
add publications from 2019 and 2020

gassmoeller committed:
Merge pull request #3520 from jdannberg/add_recent_publications

add publications from 2019 and 2020

gassmoeller committed:
Fixup changelog entry

gassmoeller committed:
Merge pull request #3519 from tjhei/fix_firefox_parameter_view

fix parameter_view in firefox

lhy11009 committed:
Merge pull request #1 from gassmoeller/phase_transition_test4

Fixup changelog entry

tjhei committed:
point values postprocessor only on rank 0

The point values postprocessor wrote the same file from all MPI ranks
potentially overwriting the results. Fix this.

ljhwang committed:
added Heyn et al. 2020 to pub list

gassmoeller committed:
Merge pull request #3496 from lhy11009/phase_transition_test4

Change input method, allowing single input value for compositions and…

gassmoeller committed:
Make \b POSIX conform

gassmoeller committed:
Fix unrecognized end

gassmoeller committed:
Update parameter files

gassmoeller committed:
Merge pull request #3522 from tjhei/point_values_postp_parallel

point values postprocessor only on rank 0

gassmoeller committed:
Merge pull request #3510 from tcclevenger/idr-solver

add IDR(s) Stokes Kylov solver

gassmoeller committed:
Update documentation

Commit:
ran make indent

anne-glerum committed:
Make pull_back_ellipsoid and push_forward_ellipsoid public

gassmoeller committed:
Merge pull request #3509 from anne-glerum/prevent_loop

Bypass topography in coordinate conversion

tjhei committed:
fix HeatingModelOutputs warning

fix clang warning

   'aspect::HeatingModel::HeatingModelOutputs' that has virtual
functions but non-virtual destructor [-Wdelete-non-virtual-dtor]

tjhei committed:
fix benchmarks and cookbooks problems

fix various warnings and errors in cookbooks and benchmarks

gassmoeller committed:
Merge pull request #3525 from tjhei/heating_model_nonvirtual

fix HeatingModelOutputs warning

gassmoeller committed:
remove special characters

naliboff committed:
Merge pull request #3500 from danieldouglas92/beam

Add Viscoelastic Cantilever Benchmark

naliboff committed:
Merge pull request #3482 from gassmoeller/extend_particle_property_interface

Extend particle property update interface

Shangxin-Liu committed:
geoid citation info

tjhei committed:
fix free-surface-with-crust

- add missing "Mesh deformation" subsection
- fix .part.prm in the manual
- fix description in manual for free-surface.prm

tjhei committed:
Merge pull request #3527 from Shangxin-Liu/geoid_citation

Add geoid citation info

naliboff committed:
Merge pull request #3528 from tjhei/free-surface-prm-fixes

fix free-surface-with-crust

anne-glerum commented on include/aspect/material_model/visco_plastic.h:
```diff
@@ -396,6 +396,13 @@ namespace aspect
         bool use_elasticity;
+
+        /**
+        * dynamic friction parameters
```
Please describe shortly what each parameter means and use capitals and punctuation as in the rest of the comments.

anne-glerum commented on source/material_model/visco_plastic.cc:
```diff
@@ -240,7 +240,12 @@ namespace aspect
           // Step 3b: calculate weakened friction, cohesion, and pre-yield viscosity
           const double current_cohesion = drucker_prager_parameters.cohesions[j] * weakening_factors[0];
-          const double current_friction = drucker_prager_parameters.angles_internal_friction[j] * weakening_factors[1];
+          // dynamic angles of internal friction must be converted from degrees into radians. static angles are converted inside drucker_prager
```
Please use capitals to start your sentences.

anne-glerum commented on source/material_model/visco_plastic.cc:
```diff
@@ -240,7 +240,12 @@ namespace aspect
           const double current_cohesion = drucker_prager_parameters.cohesions[j] * weakening_factors[0];
-          const double current_friction = drucker_prager_parameters.angles_internal_friction[j] * weakening_factors[1];
+          // dynamic angles of internal friction must be converted from degrees into radians. static angles are converted inside drucker_prager
+          // Then dynamic characterstic strain rate is used to see if dynamic or static angle of internal friction should be used
```
characterstic -> characteristic

anne-glerum commented on source/material_model/visco_plastic.cc:
```diff
@@ -240,7 +240,12 @@ namespace aspect
+          // this is done as in the material_model dynamic_friction which is based on equation 13 in van Dinther et al., (2013, JGR).
+          // const double mu  = mu_d[i] + (mu_s[i] - mu_d[i]) / ( (1 + strain_rate_dev_inv2/reference_strain_rate) );
+          // which is the following using the variables in this material_model
+          const double current_friction = dynamic_angles_of_internal_friction[j] *numbers::PI/180.0+ (drucker_prager_parameters.angles_internal_friction[j] * weakening_factors[1] - dynamic_angles_of_internal_friction[j]*numbers::PI/180.0) / (1 + std::pow((current_edot_ii/dynamic_characteristic_strain_rate[j]),dynamic_friction_smoothness_exponent[j]));
```
It's nicer if lines fit into the width of the screen of e.g. vi, ~100 characters. Please break up this line into several lines that for example start with the addition/multiplication sign. Also be consistent with spacing, e.g. one space between symbols and variable names.

anne-glerum replied:
As internally you usually always want to work with radians (this is what functions like tan and sin expect), it's clearer to convert to radians as soon as you read in the friction angles in the `parse_parameters` function. See the static angles in `rheology/drucker_prager.cc`

anne-glerum commented on source/material_model/visco_plastic.cc:
```diff
@@ -317,7 +359,8 @@ namespace aspect
               // Also convert radians to degrees
-              plastic_out->friction_angles[i] += 180.0/numbers::PI * volume_fractions[j] * (drucker_prager_parameters.angles_internal_friction[j] * weakening_factors[1]);
+              plastic_out->friction_angles[i] += 180.0/numbers::PI * volume_fractions[j] * (dynamic_angles_of_internal_friction[j] *numbers::PI/180.0+ (drucker_prager_parameters.angles_internal_friction[j] * weakening_factors[1] - dynamic_angles_of_internal_friction[j]*numbers::PI/180.0) / (1 + std::pow((current_edot_ii/dynamic_characteristic_strain_rate[j]),dynamic_friction_smoothness_exponent[j ...
+              // plastic_out->friction_angles[i] += 180.0/numbers::PI * volume_fractions[j] * (drucker_prager_parameters.angles_internal_friction[j] * weakening_factors[1]);
```
rm this line

anne-glerum commented on source/material_model/visco_plastic.cc:
```diff
@@ -317,7 +359,8 @@ namespace aspect
               plastic_out->cohesions[i]   += volume_fractions[j] * (drucker_prager_parameters.cohesions[j] * weakening_factors[0]);
               // Also convert radians to degrees
-              plastic_out->friction_angles[i] += 180.0/numbers::PI * volume_fractions[j] * (drucker_prager_parameters.angles_internal_friction[j] * weakening_factors[1]);
+              plastic_out->friction_angles[i] += 180.0/numbers::PI * volume_fractions[j] * (dynamic_angles_of_internal_friction[j] *numbers::PI/180.0+ (drucker_prager_parameters.angles_internal_friction[j] * weakening_factors[1] - dynamic_angles_of_internal_friction[j]*numbers::PI/180.0) / (1 + std::pow((current_edot_ii/dynamic_characteristic_strain_rate[j]),dynamic_friction_smoothness_exponent[j ...
```
- please break up this line into several lines
- When you have changed the internal unit of the dynamic angles, you can remove the multiple numbers::PI/180.0 here

anne-glerum commented on source/material_model/visco_plastic.cc:
```diff
@@ -726,6 +769,30 @@ namespace aspect
+          prm.declare_entry ("Dynamic characteristic strain rate", "1e-13",
+                             Patterns::Double (0),
+                             "The characteristic strain rate value, where the angle of friction takes the middle "
+                             "between the dynamic and the static angle of friction. When the effective strain rate "
```
This description sounds like a step-like change from static to dynamic friction angles, but that's not the case right?

anne-glerum commented on source/material_model/visco_plastic.cc:
```diff
@@ -726,6 +769,30 @@ namespace aspect
+          prm.declare_entry ("Dynamic angles of internal friction", "0",
+                             Patterns::List(Patterns::Double(0)),
+                             "Dynamic angles of friction which are taken when the effective strain rate in a cell "
+                             "surpasses the dynamic characteristic strain rate. "
```
see above comment

anne-glerum commented on source/material_model/visco_plastic.cc:
```diff
@@ -726,6 +769,30 @@ namespace aspect
+                             "The characteristic strain rate value, where the angle of friction takes the middle "
+                             "between the dynamic and the static angle of friction. When the effective strain rate "
+                             "in a cell surpasses the dynamic characteristic strain rate, the dynamic angle of "
+                             "friction is taken, otherwise the static angle of internal friction is chosen."
```
Add space before last ", otherwise the next line will start directly in the output, instead of after a space.

anne-glerum commented on source/material_model/visco_plastic.cc:
```diff
@@ -726,6 +769,30 @@ namespace aspect
+                             Patterns::List(Patterns::Double(0)),
+                             "An exponential factor in the equation for the calculation of the friction angle "
+                             "when a static and a dynamic friction angle is specified. A factor =1 is equivalent"
+                             " to equation 13 in van Dinther et al., (2013, JGR). A factor between 0 and 1 makes the "
```
Please remove the space after the first " and put it before the last " of the previous line.

anne-glerum commented on source/material_model/visco_plastic.cc:
```diff
@@ -849,7 +916,18 @@ namespace aspect
+                                                                                        n_fields,
+                                                                                        "Dynamic angles of internal friction");
 
+          dynamic_friction_smoothness_exponent = Utilities::possibly_extend_from_1_to_N (Utilities::string_to_double(Utilities::split_string_list(prm.get("Dynamic friction smoothness exponent"))),
```
Is it necessary that the smoothness exponent can vary between compositions?

EstherHeck replied:
true, but how do I do it?  looking at min_visc I thought this is the way, but it does not compile: 
dynamic_friction_smoothness_exponent = prm.get_double("Dynamic friction smoothness exponent");
This is the start of the error: 
/home/hecken/code/aspect/source/material_model/visco_plastic.cc:939:48: error: no match for ‘operator=’ (operand types are ‘std::vector<double, std::allocator<double> >’ and ‘double’)
           dynamic_friction_smoothness_exponent = prm.get_double("Dynamic friction smoothness exponent");

anne-glerum replied:
In the header (.h) file, the type of the parameter `dynamic_friction_smoothness_exponent` is set. At the moment, it is a vector of doubles, so you cannot assign a double to the vector as a whole. When you just want one exponent for all fields, you can change the type to `double` and keep the change you suggest (`prm.get_double`).

EstherHeck replied:
thanks!

ljhwang committed:
Added Glerum Nature Communications publication

ljhwang committed:
Added Muluneh GGG publication

tjhei committed:
test cookbook/benchmarks

- generate makefiles to configure/compile/validate all files under
cookbooks/ and benchmarks/
- optionally also execute the .prm files
- add the validation to jenkins

tjhei committed:
fix geoid warning

tjhei committed:
bencharmsk/newton_solver_benchmark_set/spiegelman bash weirdness

fix header of bash files in benchmark

related to #3535

tjhei committed:
Update benchmarks/buiter_et_al_2016_jsg/exp_1.prm

Co-authored-by: naliboff <john.naliboff@gmail.com>

gassmoeller committed:
Merge pull request #3536 from tjhei/benchmark-newton-spiegelbash

bencharmsk/newton_solver_benchmark_set/spiegelman bash weirdness

gassmoeller committed:
Merge pull request #3531 from tjhei/geoid_warn

fix geoid warning

gassmoeller committed:
Merge pull request #3523 from ljhwang/addbib_heyn

added Heyn et al. 2020 to pub list

naliboff committed:
Merge pull request #3530 from tjhei/cookbook-and-benchmark-fixes

fix benchmarks and cookbooks problems

gassmoeller committed:
Merge branch 'master' into addbib_glerum

Commit:
changelog

tjhei committed:
fix unused variable warning in register macro

Using a global variable causes an unused variable warning. Use a struct
instance with a constructor instead.

tjhei committed:
fix cookbooks/prescribed_velocity/ in 64 bit mode

with 64 bit indices, compilation fails.

tjhei committed:
remove old files and update instructions

tjhei committed:
fix two doxygen formula problems

tjhei committed:
update copyright years

tjhei committed:
gravity_point_values: only write on rank 0

fixes #3521

gassmoeller committed:
Merge pull request #3541 from tjhei/gravity_point_values_mpi

gravity_point_values: only write on rank 0

gassmoeller committed:
Merge pull request #3529 from ljhwang/addbib_glerum

Add to citating_aspect Glerum & Muluneh

Commit:
move variables

gassmoeller committed:
Merge pull request #3540 from tjhei/update_copyright

update copyright years

Commit:
fix error

gassmoeller committed:
Merge pull request #3495 from tcclevenger/gmg_q1_viscosity

add Q1 viscosity projection for GMG

gassmoeller committed:
Merge pull request #3538 from tjhei/64bit-prescribed-velocity

fix cookbooks/prescribed_velocity/ in 64 bit mode

gassmoeller committed:
Merge pull request #3539 from tjhei/doxygen-fixes

fix doxygen formula problems

tjhei committed:
avoid deprecation warning in std_cxx14 with deal.II master

tjhei committed:
avoid invalid arrary access warnings

gassmoeller committed:
Merge pull request #3537 from tjhei/benchmark-check-makefile

[CI] test cookbooks/benchmarks

anne-glerum commented on source/material_model/visco_plastic.cc:
```diff
@@ -726,6 +769,30 @@ namespace aspect
                              "0.3 $K/km$ = 0.0003 $K/m$ = 9.24e-09 $K/Pa$ gives an earth-like adiabat."
                              "Units: $K/Pa$");
 
+          //Dynamic friction paramters
```
Space before Dynamic

anne-glerum reviewed (changes requested):
Hi Esther,
Thanks for putting in the dynamic friction. My comments are mostly about formatting; so to adhere to ASPECT's general style and improve readability. Let me know if anything is unclear.

gassmoeller committed:
Merge pull request #3542 from tjhei/deal-master-warn

avoid deprecation warning in std_cxx14 with deal.II master

gassmoeller committed:
Merge pull request #3493 from MFraters/optimize_defect_correction_stokes_solver

add more tests for reference and start optimizing.

gassmoeller committed:
Merge pull request #3543 from tjhei/gcc10-warnings

avoid invalid arrary access warnings

Commit:
phase transition: average phase value

naliboff committed:
add citation

MFraters committed:
update world builder to version v0.3.0

MFraters committed:
update gwb version number in test input files.

MFraters committed:
Add changelog entry.

MFraters committed:
fix warning in some version of gcc 5.4.0

gassmoeller committed:
Merge pull request #3545 from naliboff/add_citation

Add citation

MFraters committed:
check whether separate TwoMergedBoxes model checks are needed.

MFraters committed:
adress comments John.

MFraters committed:
Turn drucker prager compositions into a lib plugin.

MFraters committed:
update input file to make it run again.

MFraters committed:
Apply suggestions from code review from John Naliboff

Co-authored-by: naliboff <john.naliboff@gmail.com>

MFraters committed:
Adress more comments from John Naliboff

tjhei committed:
support geoid dynamic citation

- add bib entry
- rename key to "geoid" from "Geoid"

tjhei committed:
use proper namespace for ZeroFunction

deal.II master removed the deprecated ZeroFunction outside the
Functions namespace.

tjhei committed:
fix for deal.II 9.1

tjhei committed:
fix prm files

tjhei committed:
remove last column in data file

tjhei committed:
fix benchmarks/yamauchi_takei*

gassmoeller committed:
Merge pull request #3550 from tjhei/zerofunction

use proper namespace for ZeroFunction

gassmoeller committed:
Merge pull request #3547 from tjhei/add_geoid_citation

support geoid dynamic citation

gassmoeller committed:
Merge pull request #3553 from tjhei/benchmark_fixes1

fix  benchmarks/yamauchi_takei_2016_anelasticity

tjhei committed:
fix several benchmarks

fix benchmarks/slab_detachment/ data directory
only run specific prms in rayleigh_taylor_instability and sinking_block

gassmoeller committed:
Merge pull request #3546 from MFraters/update_world_builder_version_to_v0.3.0

Update world builder version to v0.3.0

tjhei committed:
clang-tidy fixes

tjhei committed:
add GWB and Newton to dynamic citation page

MFraters committed:
Merge pull request #3555 from tjhei/clang-tidy-fixes

clang-tidy fixes

tjhei committed:
remove duplicate

Commit:
changed first formatting comments from Anne

Commit:
ran make indent

Commit:
changed the default of the dynamic friction angles to be the internal friction angles

Commit:
ran make indent

Commit:
inserted comments about dynamic friction parameters

Commit:
smoothness exponent is now only one field, not as many as materials any more

gassmoeller committed:
Merge pull request #3554 from tjhei/benchmark_fixes3

fix several benchmarks

tjhei committed:
change FBTGS18 to FBTGS19

MFraters committed:
Merge pull request #3557 from tjhei/cite_wb

add GWB and Newton to dynamic citation page

gassmoeller committed:
Merge pull request #3183 from MFraters/add_more_spiegelman_benchmark_files

Add more spiegelman benchmark files

gassmoeller committed:
Add Spiegelman newton solver benchmark check

gassmoeller committed:
Merge pull request #3552 from tjhei/fix_dap_example

Fix opendap example

tjhei committed:
Merge pull request #3559 from gassmoeller/add_spiegelman_benchmark_check

Add Spiegelman newton solver benchmark check

tjhei committed:
update link in readme

gassmoeller committed:
Update substitute pattern

tjhei committed:
make_cite_html.py fix formatting in citation

tjhei committed:
update copyright

tjhei committed:
Merge pull request #3458 from gassmoeller/make_update_script_standard_compliant

Make parameter file update script work on Mac

gassmoeller committed:
Merge pull request #3562 from tjhei/readme_contrib

update link in readme

gassmoeller committed:
Merge pull request #3564 from tjhei/copyright2

update copyright

gassmoeller committed:
Merge pull request #3563 from tjhei/cite_wb_2

make_cite_html.py fix formatting in citation

tjhei committed:
fix benchmarks/viscoelastic_beam_modified

this fixes the error
Boundary indicator <2> with symbolic name <bottom> is listed as having more than one type of velocity or traction boundary condition in
the input file.

naliboff committed:
Merge pull request #3565 from tjhei/fix_benchmark_viscoelastic_beam_modified

fix benchmarks/viscoelastic_beam_modified

tjhei committed:
release task: update version and changes.h

Commit:
internal angles of friction now default for dynamic angles of friction. ran make indent

tjhei committed:
add 2.2 announcement

gassmoeller committed:
Merge pull request #3566 from tjhei/post-release-2.2.0

release task: update version and changes.h

gassmoeller committed:
Update release announcement

gassmoeller committed:
Merge announcements

cedrict committed:
adding the two missing files of the viscosity_grooves benchmark

gassmoeller committed:
Address comments

tjhei committed:
Merge pull request #10 from gassmoeller/release_announcement_2.2.0

Release announcement 2.2.0

tjhei committed:
fix name

tjhei committed:
Merge pull request #3568 from cedrict/readd_files_visc_groove

adding the two missing files of the viscosity_grooves benchmark

tjhei committed:
update viscosity_grooves benchmark

- use double numbers
- update copyright years
- use compute_global_error
- enable in automatic check

gassmoeller committed:
Merge pull request #3567 from tjhei/announcement

add 2.2 announcement

tjhei committed:
Merge pull request #3569 from tjhei/viscosity_grooves_update

update viscosity_grooves benchmark

tjhei committed:
FIND_PACKAGE bump required ASPECT version

- change required ASPECT version to 2.2 (even though the version on
master is 2.3.0-pre now)
- also bump the deal.II version that was forgotten
- add some missing license headers

gassmoeller committed:
Merge pull request #3570 from tjhei/bump_cmake_aspect_version

FIND_PACKAGE bump required ASPECT version

tjhei committed:
update DOI in README

tjhei committed:
update citing.html and bib entries

- update zenodo/figshare links in bib file
- update citing.html database with new entries

gassmoeller committed:
Merge pull request #3574 from tjhei/release-readme

update zenodo DOI in README

gassmoeller committed:
Merge pull request #3576 from tjhei/cite_html_update

update citing.html and bib entries

Commit:
inserted check for the dynamic friction angles to be <= 90.

tjhei committed:
update release task list

tjhei committed:
update parameters

July maintenance

gassmoeller committed:
Merge pull request #3581 from tjhei/release_task_list

update release task list

lhy11009 committed:
Update doc/modules/changes/20200701_haoyuanli

Co-authored-by: Rene Gassmoeller <rene.gassmoeller@mailbox.org>

tjhei committed:
remove spack yaml file

this file doesn't do anything useful, so remove it. See #3577

tjhei committed:
introduce config.h for compile definitions

this removes the need for COMPILE_DEFINITIONS, which interferes with
UNITY_BUILD.

see #3585

tjhei committed:
make plugins work

tjhei committed:
fix world builder define

tjhei committed:
undef INSTANTIATE macro

this is needed for unity builds to work correctly

tjhei committed:
unity build fixes

Commit:
ran make indent

gassmoeller committed:
Merge pull request #3584 from tjhei/maintenance_july2020

update parameters

gassmoeller committed:
Merge pull request #3588 from tjhei/remove-spack

remove spack yaml file

gassmoeller committed:
Merge pull request #3544 from lhy11009/phase_transition_test6

phase transition: average phase value

tjhei committed:
remove cotire macro undef

gassmoeller committed:
Merge pull request #3589 from tjhei/compile_definitions_in_config_h

introduce config.h for compile definitions

gassmoeller committed:
Style fixes.

gassmoeller committed:
Merge pull request #3591 from tjhei/unity-fixes

unity build fixes

gassmoeller committed:
Merge pull request #3590 from tjhei/undef_instantiate_macro

undef INSTANTIATE macro

gassmoeller committed:
Merge pull request #3488 from MFraters/fix_two_merged_boxes

separate two merged boxes from box geometry model

gassmoeller committed:
Use cmake to precompile headers if possible

gassmoeller committed:
Remove unnecessary include from simulator_access.h

tjhei committed:
Merge pull request #3592 from gassmoeller/avoid_unnecessary_includes

Remove unnecessary include from simulator_access.h

tjhei committed:
disable world builder in cmake UNITY_BUILD

gassmoeller committed:
Move variable declaration to more reasonable location

gassmoeller committed:
Merge pull request #3593 from tjhei/unity_no_wb

disable world builder in cmake UNITY_BUILD

gassmoeller committed:
Disable precompiling headers unless supported.

tjhei committed:
Merge pull request #3587 from gassmoeller/use_cmake_precompile_headers

Use native cmake to precompile headers

Commit:
added description at the end of .cc file and ran make indent
